doc/developer/design/20230531_system_privileges.md

```diff
@@ -0,0 +1,97 @@
+# System Privileges
+
+- Associated: [RBAC](20230216_role_based_access_control.md)
+
+## Context
+
+Role based access control comprises a set of features that allow customers to control the
+privileges of individual users. Object privileges control user's access to database objects, they
+are granted with the `GRANT` keyword and revoked with the `REVOKE` keyword. For example,
+`GRANT SELECT ON TABLE t TO r` will grant the ability to read table `t` to role `r`. `GRANT CREATE
+ON SCHEMA s TO r` will grant the ability to create new objects in schema `s` to role `r`. Object
+privileges can be inherited through role membership. Role attributes control the user's ability to
+create (and sometimes modify) top level objects, they are granted and revoked with the `ALTER ROLE`
+keywords. For example, `ALTER ROLE r CREATEDB` will grant the ability to create databases to
+role `r`. `ALTER ROLE r CREATECLUSTER` will grant the ability to create clusters to role `r`. Role
+attributes cannot be inherited through role membership.
+
+Role attributes always exist in a boolean state. To change the state of a role attribute you must
+alter the role. For example `ALTER ROLE joe CREATEDB` sets the `CREATEDB` attribute to `true`,
+and `ALTER ROLE NOCREATEDB` sets the `CREATEDB` attribute to `false`.
+
+On the other hand, privileges either exist or they don't exist. You create them by granting a
+privilege via `GRANT ...` and you destroy them via `REVOKE ...`.
+
+Role attributes present some problems as they currently exist:
+
+- It can be confusing that there are two separate privilege primitives that control a role's
+  privilege, and they have different semantics.
+- Role attributes are not inherited which makes it cumbersome to grant attributes to a large
+  number of roles.
+- Users are not able to easily grant the ability to create top level objects to many current and
+  future users in a short amount of time.
+
+## Goals
+
+- Reduce confusion around privileges.
+- Allow the ability to create databases, clusters, and roles to be inherited through role
+  membership.
+- Allow users to easily grant the ability to create top level objects to many current and future
+  users in a short amount of time.
+- Maintain PostgreSQL compatibility.
+
+## Overview
+
+We will add a new type of privilege called System Privileges. They will look and act just like
+normal object privileges, but they will grant the abilities that would otherwise be granted through
+role attributes. They will exist in parallel with role attributes.
+
+NOTE: Other databases, such as CockroachDB and Snowflake, have some form of system privileges that
+control similar privileges to role attributes.
```

Comment:
Do these databases also support Role Attributes?

I'm not opposed to having both Role Attributes and System Privileges, but System Privileges seem strictly better, so I'm wondering if we can just get rid of roles, or retain the Roles API, but have it be a shim on top of system privileges

Reply:
CockroachDB has both role options (which is what they call role attributes) and system privileges, though they have marked role options as deprecated. As far as I know, Snowflake does not have role attributes, but they do not have any PG compatibility concerns.

We could implement role attributes as a "view" over system privileges, but then role attributes would be inheritable, which would break part of the PG compatibility.

I'm still not opposed to breaking PG compatibility here and just getting rid of role attributes if people think the trade-off is worth it.